### PR TITLE
Implemented inference in template plugin

### PR DESCRIPTION
**Pull request by apankratovantonp**
* Added inference using `ngraph Interpreter` back end
* Added `QueryNetwork` API implementation
* Added Threading Configuration

### CODE REVIEW

File: docs/template_plugin/src/CMakeLists.txt [hunk 1]
```diff
@@ -23,11 +23,18 @@ target_include_directories(${TARGET_NAME} PRIVATE
...
     "${IE_MAIN_TEMPLATE_PLUGIN_SOURCE_DIR}/include")
 
-target_link_libraries(${TARGET_NAME} PRIVATE IE::inference_engine IE::inference_engine_transformations ${NGRAPH_LIBRARIES} ${INTEL_ITT_LIBS})
+target_link_libraries(${TARGET_NAME} PRIVATE
+    IE::inference_engine
+    IE::inference_engine_transformations
+    ${INTEL_ITT_LIBS}
+    ${NGRAPH_LIBRARIES}
+    ngraph_backend
+    interpreter_backend
```
Comment (ilyachur) [inline docs/template_plugin/src/CMakeLists.txt:32]: Do we really want to link template plugin with some libraries for tests? Maybe should we add comments in order to warn user that it is not required component and he shouldn't use it in his own plugin.

File: docs/template_plugin/src/CMakeLists.txt [hunk 2]
```diff
@@ -23,11 +23,18 @@ target_include_directories(${TARGET_NAME} PRIVATE
...
+    half)
 
 # ATTENTION: uncomment to register a plugin in the plugins.xml file
-# ie_register_plugins(MAIN_TARGET ${TARGET_NAME}
-                    # POSSIBLE_PLUGINS ${TARGET_NAME})
+ie_register_plugins(MAIN_TARGET ${TARGET_NAME}
```
Comment (ilyachur) [inline docs/template_plugin/src/CMakeLists.txt:36]: Do we remove template_plugin from `plugins.xml` when prepare a OpenVINO package?

File: inference-engine/thirdparty/CMakeLists.txt [hunk 3]
```diff
@@ -53,12 +53,28 @@ function(build_with_lto)
 
     target_include_directories(pugixml INTERFACE "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/pugixml/src>")
 
+    add_library(half INTERFACE)
```
Comment (ilyachur) [inline inference-engine/thirdparty/CMakeLists.txt:56]: Do we add a new third party library only for template_plugin?

File: docs/template_plugin/src/CMakeLists.txt (hunk 1 above)
Comment (ilyachur) [reply docs/template_plugin/src/CMakeLists.txt:32]: BTW `interpreter_backend` supports only limited scope of operations.

File: inference-engine/thirdparty/CMakeLists.txt (hunk 3 above)
Comment (apankratovantonp) [reply inference-engine/thirdparty/CMakeLists.txt:56]: * I have another plugin that can use it. 
* I can move it to third party directory inside plugin folder
* I believe that all host type conversions should be done with GAPI preprocessing here. But it is not implemented yet.

File: docs/template_plugin/src/CMakeLists.txt (hunk 1 above)
Comment (apankratovantonp) [reply docs/template_plugin/src/CMakeLists.txt:32]: Moved to separate section. Should not `interpreter_backend` supports all operations from current opset as reference code as it is used for tests?

File: inference-engine/thirdparty/CMakeLists.txt (hunk 3 above)
Comment (apankratovantonp) [reply inference-engine/thirdparty/CMakeLists.txt:56]: Removed

Comment (apankratovantonp) [reply inference-engine/thirdparty/CMakeLists.txt:56]: @ilyachur  BTW Do we have ability to convert fp16 ngraph::function into fp32 as it is done in CPU plugin but using ngrap only?

Comment (ilyachur) [reply inference-engine/thirdparty/CMakeLists.txt]: @apankratovantonp At the current moment we don't have transformation pass to convert fp16 to fp32... 
Some time ago we added only path for fp32 to fp16 conversion (ngraph/pass/convert_fp32_to_fp16.hpp)

File: docs/template_plugin/src/CMakeLists.txt (hunk 1 above)
Comment (ilyachur) [reply docs/template_plugin/src/CMakeLists.txt:37]: At the current moment Interpret backend supports only operations which have reference implementations. The scope of supported operation is extending (it is needed for tests) but coverage is not 100%.

BBTW why template plugin should support to infer all operations form the last opset?

Comment (apankratovantonp) [reply docs/template_plugin/src/CMakeLists.txt:37]: Why not?

File: docs/template_plugin/src/CMakeLists.txt [hunk 4]
```diff
@@ -20,14 +20,25 @@ ie_add_plugin(NAME ${TARGET_NAME}
               VERSION_DEFINES_FOR template_plugin.cpp)
 
 target_include_directories(${TARGET_NAME} PRIVATE
-    "${CMAKE_CURRENT_SOURCE_DIR}"
+    "${CMAKE_CURRENT_SOURCE_DIR}")
+
+target_include_directories(${TARGET_NAME} PUBLIC
```
Comment (ilya-lavrenov) [inline docs/template_plugin/src/CMakeLists.txt:25]: why `PUBLIC`? `Template` plugin cannot be linked and so `PUBLIC` is not needed

File: docs/template_plugin/src/template_infer_request.hpp [hunk 5]
```diff
@@ -57,17 +61,18 @@ class TemplateInferRequest : public InferenceEngine::InferRequestInternal {
...
 
-    // for performance counts
-    double                                                  _inputPreprocessTime   = 0.0;
-    double                                                  _inputTransferTime     = 0.0;
-    double                                                  _executeTime           = 0.0;
-    double                                                  _outputTransferTime    = 0.0;
-    double                                                  _outputPostProcessTime = 0.0;
+    std::vector<std::shared_ptr<ngraph::runtime::Tensor>>   _inputTensors;
+    std::vector<std::shared_ptr<ngraph::runtime::Tensor>>   _outputTensors;
+    std::shared_ptr<ngraph::runtime::Executable>            _executable;
```
Comment (ilya-lavrenov) [inline docs/template_plugin/src/template_infer_request.hpp:75]: should not it be created per executable network?

File: docs/template_plugin/src/template_infer_request.cpp [hunk 6]
```diff
@@ -65,159 +66,180 @@ void TemplateInferRequest::allocateDeviceBuffers() {
...
+        networkBlobMap[blobData.first] = networkBlob;
     }
+}
 
-    if (_networkOutputs.empty() || _networkInputs.empty()) {
-        THROW_IE_EXCEPTION << "Internal error: no information about network's output/input";
-    }
+void TemplateInferRequest::allocateBlobs() {
+    _inputTensors.resize(_networkInputs.size());
+    _outputTensors.resize(_networkOutputs.size());
```
Comment (ilya-lavrenov) [inline docs/template_plugin/src/template_infer_request.cpp:112]: `_outputTensors` and `_inputTensors` look like "device buffers" in current context. Should we move their creation to `allocateDeviceBuffers`?

File: inference-engine/thirdparty/CMakeLists.txt (hunk 3 above)
Comment (ilya-lavrenov) [reply inference-engine/thirdparty/CMakeLists.txt]: `half` library is removed

**Comment by ilya-lavrenov** [comment]
@apankratovantonp let's make the CI green and take into account 2 latest comments.

File: docs/template_plugin/tests/functional/transformations/template_transformations_test.cpp [hunk 7]
```diff
@@ -18,7 +18,7 @@
 using namespace testing;
 
 // ! [transformation:test]
-TEST(TransformationTests, TemplateTest) {
+TEST(TransformationTests, DISABLED_TemplateTest) {
```
Comment (ilya-lavrenov) [inline docs/template_plugin/tests/functional/transformations/template_transformations_test.cpp:21]: why this test is disabled? Does not it pass even in master?

File: docs/template_plugin/src/template_executable_network.cpp [hunk 8]
```diff
@@ -16,93 +16,85 @@
...
 // ! [executable_network:ctor_cnnnetwork]
-TemplatePlugin::ExecutableNetwork::ExecutableNetwork(InferenceEngine::ICNNNetwork&  network,
-                                                     const Configuration&           cfg):
-    _name(network.getName()),
+TemplatePlugin::ExecutableNetwork::ExecutableNetwork(const std::shared_ptr<ngraph::Function>&   function,
+                                                     const Configuration&                       cfg,
+                                                     const Plugin::Ptr&                         plugin) :
+    InferenceEngine::ExecutableNetworkThreadSafeDefault(nullptr, nullptr), // Disable default threads creation
     _cfg(cfg),
-    _waitExecutor(InferenceEngine::ExecutorManager::getInstance()->getExecutor("Template")) {
```
Comment (ilya-lavrenov) [inline docs/template_plugin/src/template_executable_network.cpp:34]: why `_waitExecutor` is moved out of executable network?

File: docs/template_plugin/src/template_infer_request.hpp (hunk 5 above)
Comment (apankratovantonp) [reply docs/template_plugin/src/template_infer_request.hpp:75]: I am sure that it is not thread safe to use it in multi threaded environment so it is better to have a copy in each infer request.

File: docs/template_plugin/src/template_executable_network.cpp (hunk 8 above)
Comment (apankratovantonp) [reply docs/template_plugin/src/template_executable_network.cpp:34]: The Idea is to have one executor per device for all executable networks in plugin. So plugin is better place for it

File: docs/template_plugin/tests/functional/transformations/template_transformations_test.cpp (hunk 7 above)
Comment (apankratovantonp) [reply docs/template_plugin/tests/functional/transformations/template_transformations_test.cpp:21]: It dose not work in master

File: docs/template_plugin/src/template_infer_request.cpp (hunk 6 above)
Comment (apankratovantonp) [reply docs/template_plugin/src/template_infer_request.cpp:112]: moved

**Comment by ilya-lavrenov** [comment]
@apankratovantonp please, make CI green and I will merge

**Pull request merged**